prov/rstream/src/rstream_ep.c

```diff
@@ -0,0 +1,343 @@
+#include "rstream.h"
+
+static int rstream_ep_close(fid_t fid)
+{
+    int ret;
+    struct rstream_ep *rstream_ep =
+        container_of(fid, struct rstream_ep, util_ep.ep_fid.fid);
+
+    ret = fi_close(&rstream_ep->local_mr.mr->fid);
+    if (ret)
+        return ret;
+    free(rstream_ep->local_mr.base_addr);
+
+    ret = fi_close(&rstream_ep->ep_fd->fid);
+    if (ret)
+        return ret;
+
+    ret = fi_close(&rstream_ep->recv_cq->fid);
+    if (ret)
+        return ret;
+
+    ret = fi_close(&rstream_ep->send_cq->fid);
+    if (ret)
+        return ret;
+
+    ofi_endpoint_close(&rstream_ep->util_ep);
+    free(rstream_ep->tx_ctx.tx_ctxs);
+    free(rstream_ep->rx_ctxs);
+    free(rstream_ep);
+    return 0;
+}
+
+static int rstream_ep_bind(struct fid *ep_fid, struct fid *bfid, uint64_t flags)
+{
+    int ret;
+    struct rstream_ep *rstream_ep =
+        container_of(ep_fid, struct rstream_ep, util_ep.ep_fid.fid);
+    struct rstream_eq *rstream_eq = NULL;
+
+    switch (bfid->fclass) {
+    case FI_CLASS_EQ:
+        rstream_eq = container_of(bfid, struct rstream_eq, util_eq.eq_fid.fid);
+        ret = fi_ep_bind(rstream_ep->ep_fd, &rstream_eq->eq_fd->fid, flags);
+        /* latest bind will always be associated with next cm xchg
+        given rsockets impl. (aka bind then either connect/accept */
+        rstream_eq->bound_ep = rstream_ep;
+	    break;
+    default:
+        FI_WARN(&rstream_prov, FI_LOG_EP_CTRL, "invalid fid class\n");
+        ret = -FI_EINVAL;
+        break;
+    }
+    return ret;
+}
+
+static int rstream_ep_ctrl(struct fid *fid, int command, void *arg)
+{
+    struct rstream_ep *rstream_ep;
+    int ret;
+    rstream_ep = container_of(fid, struct rstream_ep, util_ep.ep_fid.fid);
+
+    switch (command) {
+    case FI_ENABLE:
+        ret = fi_enable(rstream_ep->ep_fd);
+        break;
+    case FI_GETWAIT: /* need to have one cq->fd */
```

Comment:
You will need to use a single CQ or create a wait set. A single CQ should be a better option.

Reply:
this is the aspect that is a work in progress pending the poll implementation; I agree that those are the options, I'm just not quite there yet feature set wise